Add missing trusted cluster info to TLS certs created at tsh login

awly opened the pull request:
TLS credentials on login appear different from them on re-issue.
Specifically, they are missing kube users/groups in Subject, as well as
`RouteToCluster`.

With this change, user can now do:
```
$ tsh login --proxy=root-proxy leaf-proxy
$ kubectl get ...
```
and kubectl commands will get routed through `root-proxy` to `leaf-proxy`
and contain proper kube principals for the user.

Another fix bundled here is advertising `kube.listen_addr` via proxy `Ping` RPC, along with existing `kube.public_addr`. If `listen_addr` used a custom port, the client wouldn't figure it out.

Fixes #3693

Commit:
Advertise k8s listen_addr from the proxy

If kube.public_addr is not set and kube.listen_addr uses a non-standard
port, the client can't discover it. Advertise kube.listen_addr same as
we advertise it for SSH.

Also, override client.TeleportClient.KubeProxyAddr with info from
proxy's Ping response. Existing value in that field comes from
~/.tsh/profile and can contain the wrong value.

Commit:
Refactor SSH login funcs in lib/client/weblogin.go

There are 3 login flows: direct, SSO and U2F.
Only SSO had a struct grouping all the parameters. The other two had a
bunch of func arguments.

Make them consistent, use structs for arguments for all 3.

Commit:
Plumb RouteToCluster and k8s principals on login

TLS credentials on login appear different from them on re-issue.
Specifically, they are missing kube users/groups in Subject, as well as
RouteToCluster.

With this change, user can now do:
$ tsh login --proxy=root-proxy leaf-proxy
$ kubectl get ...
and kubectl commands will get routed through root-proxy to leaf-proxy
and contain proper kube principals for the user.

Commit:
Allow TLS cert creation without kube users/groups

This is a reasonable situation, users are not required to have kube
principals. Skip that error on login.

Commit:
Add a unit test for auth.AuthenticateSSHUser

This method should correctly populate k8s principals and RouteToCluster.

awly commented:
Reviewers: happy to chat over zoom if there's too much going on in this PR.

webvictim commented on lib/client/weblogin.go:
```diff
@@ -157,6 +164,62 @@ type SSHLogin struct {
+
+// SSHLoginU2F contains SSH login parameters for U2F login.
+type SSHLoginU2F struct {
+	// ProxyAddr is the target proxy address
+	ProxyAddr string
+	// User is the login username.
+	User string
+	// User is the login password.
+	Password string
+	// PubKey is SSH public key to sign
+	PubKey []byte
+	// TTL is requested TTL of the client certificates
+	TTL time.Duration
+	// Insecure turns off verification for x509 target proxy
+	Insecure bool
+	// Pool is x509 cert pool to use for server certifcate verification
+	Pool *x509.CertPool
+	// Compatibility sets compatibility mode for SSH certificates
+	Compatibility string
+	// RouteToCluster is an optional cluster name to route the response
+	// credentials to.
+	RouteToCluster string
```
As this set of values is common to all `SSHLogin*` structs, is there any way for us to define it once and reuse it? Feels like a lot of boilerplate being copy/pasted with the same comments.

awly replied:
Good idea, done

webvictim commented on lib/client/weblogin.go:
```diff
@@ -434,22 +501,23 @@ func SSHAgentLogin(ctx context.Context, proxyAddr string, user string, password
 // temporary certificate back.
-func SSHAgentU2FLogin(ctx context.Context, proxyAddr, user, password string, pubKey []byte, ttl time.Duration, insecure bool, pool *x509.CertPool, compatibility string) (*auth.SSHLoginResponse, error) {
-	clt, _, err := initClient(proxyAddr, insecure, pool)
+//func SSHAgentU2FLogin(ctx context.Context, proxyAddr, user, password string, pubKey []byte, ttl time.Duration, insecure bool, pool *x509.CertPool, compatibility string) (*auth.SSHLoginResponse, error) {
```
comment left in here

awly replied:
Ooops, removed.

webvictim reviewed (commented):
LGTM on the whole. Very happy to see more standardisation going on.

Commit:
Extract common fields of various client.SSHLogin* structs

A bit of deduplication, no functional changes.

russjones commented on lib/web/apiserver.go:
```diff
@@ -1090,7 +1090,7 @@ func (h *Handler) createSession(w http.ResponseWriter, r *http.Request, p httpro
 	}
 	if err != nil {
-		return nil, trace.AccessDenied("bad auth credentials")
+		return nil, trace.AccessDenied("bad auth credentials: %v", err)
```
Can the error here contain any sensitive information and is this error returned to the client? I'm inclined to log it instead of returning it in an error. That we we can be sure we're not accidentally returning sensitive information to the client.

awly replied:
Good point. I changed this because debugging why login failed was impossible otherwise.
Changed to logging the error, it'll work for debugging too.

russjones commented on lib/auth/methods.go:
```diff
@@ -216,6 +216,7 @@ type AuthenticateSSHRequest struct {
 	TTL time.Duration `json:"ttl"`
 	// CompatibilityMode sets certificate compatibility mode with old SSH clients
 	CompatibilityMode string `json:"compatibility_mode"`
+	RouteToCluster    string `json:"route_to_cluster"`
```
Question regarding migration, is it okay for this value to be empty? If you have an old client connecting to a new server, this value will not be set.

awly replied:
Yes, if this field is empty it will not be populated in the TLS cert (in the `street` field).
Before this change, it was never populated.
Confirmed locally with newer server and older client.

webvictim reviewed (approved):
LGTM pending @russjones' comments

Commit:
Log auth failure error instead of returning it

This prevents accidental leakage of sensitive information to the user.
The error is still logged to help debugging.

russjones reviewed (approved):
Looks good to me. Regarding the question about backporting we discussed earlier, let's hold off on backporting this since we're so close to 4.3 release. However we'll keep it as an option if we absolutely have to.
